cheatsheets/Nodejs_security_cheat_sheet.md

```diff
@@ -0,0 +1,371 @@
+    process.exit(); // exit the process to avoid unknown state
+});
+```
+
+## Monitor the event loop
+
+When your application server is under heavy network traffic, it may not be able to serve its users. This is essentially a type of Denial of Service (DoS) attack. Toobusy module allows you to monitor the event loop. It keeps track of lags and when it goes beyond a certain threshold, this module can indicate your server is too busy. In that case, you can stop processing incoming requests and send them 503 Server Too Busy message so that your application stay responsive. Sample use of toobusy module is shown here:
+
+```JavaScript
+var toobusy = require(‘toobusy’);
+var express = require(‘express’);
+var app = express();
+app.use(function(req, res, next) {
+    if (toobusy()) {
+        // log if you see necessary
+        res.send(503, “Server Too Busy”);
+    } else {
+    next();
+    }
+});
+```
+
+## Perform application activity logging
+
+Logging application activity is an encouraged good practice. It makes it easier to debug any errors encountered during application runtime. It is also useful for security concerns, since it can be used during incident response. Also, these logs can be used to feed Intrusion Detection/Prevention Systems (IDS/IPS). In Node.js, there are some modules like Winston or Bunyan to perform application activity logging. These modules enable streaming and querying logs. Also, they provide a way to handle uncaught exceptions. With the following code, you can log application activities in both console and a desired log file.
+
+```JavaScript
+var logger = new (Winston.Logger) ({
+    transports: [
+        new (winston.transports.Console)(),
+        new (winston.transports.File)({ filename: ‘application.log’ })
+    ],
+    level: ‘verbose’
+});
+```
+
+Also, you can provide different transports so that you can save errors to a separate log file and general application logs to a different log file.
+
+## Perform input validation
+
+Input validation is a crucial part of application security. Input validation failures can result in many different types of application attacks. These include SQL Injection, Cross-Site Scripting, Command Injection, Local/Remote File Inclusion, Denial of Service, Directory Traversal, LDAP Injection and many other injection attacks. In order to avoid these attacks, input to your application should be sanitized first. The best input validation technique is to use a white list of accepted inputs. However, if this is not possible, input should be first checked against expected input scheme and dangerous inputs should be escaped. In order to ease input validation in Node.js applications, there are some modules like validator and mongo-express-sanitize. Also, you should escape all HTML and JavaScript content shown to users via application. You can use [escape-html](https://github.com/component/escape-html) or [node-esapi](https://github.com/ESAPI/node-esapi) libraries to perform output escaping.
```

Comment:
Link to input validation cheatsheet.

Should output encoding be its own section?

Reply:
Added the link and create a new section for output escaping